### ISSUE TITLE
Let me edit the patch before staging or otherwise applying

### ISSUE THREAD

**Issue by Toxaris** [magit/magit#2718]
When I have edited a file and want to create a sequence of commits from my edits, I sometimes need to manually edit the diff of some of the commits. On the command line, I would use `git add -p`, then <kbd>e</kbd>. I would like to do something similar with Magit.

Maybe I'm just missing a way to achieve this with Magit as-is? I'll include a sketch of the wished-for user interface and two examples below to make it clearer what I want to achieve.

**Proposed user interface:** For every command that can operate on the hunk under point (for example, `magit-stage`), I want a variant (say, `magit-stage-after-edit`) that displays a buffer with the patch. After editing the patch in the buffer, I want to press <kbd>C-c</kbd> <kbd>C-c</kbd> to do the original action, but with the edited patch. If the patch applies, the buffer should be killed. If the patch doesn't apply, the buffer should stay and an error message should be shown somewhere, so that I have a chance of correcting my patch and trying <kbd>C-c</kbd> <kbd>C-c</kbd> again. If I press <kbd>C-c</kbd> <kbd>C-k</kbd>, the buffer should be killed and the operation should be aborted. In particular, <kbd>M-x</kbd> `magit-stage-after-edit` <kbd>C-c</kbd> <kbd>C-c</kbd> should have the same overall effect as just <kbd>M-x</kbd> `magit-stage`.

**First example:** When these are my overall changes to my emacs config:

``` diff
-(use-package magit)
+(use-package magit
+  :ensure t
+  :config
+  (global-set-key (kbd "C-x g") 'magit-status))
```

I want to split this into two commits, one for `:ensure` and one for `:config`. So I want to create a first commit ("Install magit automatically"):

``` diff
-(use-package magit)
+(use-package magit
+  :ensure t)
```

And a second commit ("Bind C-x g to start magit.")

``` diff
 (use-package magit
-  :ensure t)
+  :ensure t
+  :config
+  (global-set-key (kbd "C-x g") 'magit-status))
```

**Second example:** When these are my overall changes to a LaTeX document:

``` diff
-\label{fig:foo}
+\label{fig:bar}
 \end{figure}

-The code in \autoref{fig:foo} was written by the authors
+We wrote the code in \autoref{fig:bar}.
```

I want to split the changes into two commits, one for changing the voice, and one for the refactoring. So I want to produce a first commit ("Avoid passive voice"):

``` diff
-The code in \autoref{fig:foo} was written by the authors
+We wrote the code in \autoref{fig:foo}.
```

And a second commit ("Rename fig:foo to fig:bar."):

``` diff
-\label{fig:foo}
+\label{fig:bar}
 \end{figure}

-We wrote the code in \autoref{fig:foo}.
+We wrote the code in \autoref{fig:bar}.
```

**Comment by npostavs** [comment]
For separating lines in hunks you can use magit's partial hunk (un)staging/applying.

For picking apart intraline changes, I think `E s` (`magit-ediff-stage`) works better than editing the diff directly.

https://magit.vc/manual/magit/Staging-and-unstaging.html

**Comment by Toxaris** [comment]
Thanks for pointing me to `magit-ediff-stage`. I've never used it before, and the documentation in the magit manual is rather sparse, so I didn't get that it might be useful in my situation.

If I understand correctly, `magit-ediff-stage` let's me do a three-way merge between HEAD, the index, and the working copy. So for staging, I would manually modify the index to the state I want it to be and then quit the ediff session? (This doesn't appear to be explained in the magit manual)

It's not clear to me why ediff should work better than editing the diff for separating intraline changes. After all, all other operations in magit are on diffs, and the commit will later be reviewed as a diff, so my mind is pretty much in diff-mode when commiting. However, since I cannot (yet) operate ediff, I'll have to give it a try sometime to find out.

**Comment by kyleam** [comment]
Related to #2173.

**Comment by npostavs** [comment]
> I've never used it before, and the documentation in the magit manual is rather sparse,

Yeah, I didn't really get how it worked until fixing #2701.

> If I understand correctly, `magit-ediff-stage` let's me do a three-way merge between HEAD, the index, and the working copy. So for staging, I would manually modify the index to the state I want it to be and then quit the ediff session?
> 
> It's not clear to me why ediff should work better than editing the diff for separating intraline changes.

You can edit the text that will be staged directly, instead of editing the text of the diff that will modify what will be staged. One less layer of indirection.

**Comment by npostavs** [comment]
Here's a proof-of-concept, using `recursive-edit` (you need to hit <kbd>C-M-c</kbd> rather than <kbd>C-c C-c</kbd>, <kbd>C-]</kbd> to abort). I still ediff gives a better interface.

```diff
diff --git i/lisp/magit-apply.el w/lisp/magit-apply.el
index a99000a..e4e998e 100644
--- i/lisp/magit-apply.el
+++ w/lisp/magit-apply.el
@@ -148,6 +148,14 @@ (defun magit-apply-region (section &rest args)
                      (concat (magit-diff-file-header section)
                              (magit-diff-hunk-region-patch section args))))
 
+(defvar magit-edit-patch nil
+  "Let-bind to non-nil, to edit patch before applying.")
+
+(defun magit-stage-after-edit ()
+  (interactive)
+  (let ((magit-edit-patch t))
+    (call-interactively 'magit-stage)))
+
 (defun magit-apply-patch (section:s args patch)
   (let* ((files (if (atom section:s)
                     (list (magit-section-value section:s))
@@ -160,6 +168,13 @@ (defun magit-apply-patch (section:s args patch)
       (magit-wip-commit-before-change files (concat " before " command)))
     (with-temp-buffer
       (insert patch)
+      (when magit-edit-patch
+        ;; Temp buffers don't have font lock by default,
+        ;; so we need to rename to make it usable for editing.
+        (rename-buffer "*Edit Magit Patch*" t)
+        (diff-mode)
+        (switch-to-buffer (current-buffer))
+        (recursive-edit))
       (magit-run-git-with-input
        "apply" args "-p0"
        (unless (magit-diff-context-p) "--unidiff-zero")
```

**Comment by tarsius** [comment]
Instead of learning about Ediff and then using that you could also just edit the *file*, stage and commit, and then undo the changes.

This would involve the following steps:

1. Use <kbd>RET</kbd> on the file or hunk to go to the file-visiting buffer. Point is automatically placed on the change. Enabling `diff-hl-mode` would help seeing what changed.
2. Make the change and save.
3. Go back to the status (or diff) buffer using <kbd>C-x b RET</kbd>.
4. Stage and then commit.
5. Visit the file-visiting buffer.
6. Undo the changes.
7. Go back to status buffer.
8. Stage and commit.

These are quite a few steps, but I doubt that the Ediff approach would take much less key strokes.

---

If we do implement support for editing a diff before applying it, then I think this should be done using two commands. The first should populate and visit a `diff-mode` buffer and somehow (likely using a minor-mode) bind a key to the second command which would feed the buffer content to `magit-apply-patch` (which would have to be taught to except `nil` as SECTION:S).


Using these new commands would involve the following steps:

1. Use first command.
2. Edit.
3. Use second command.
4. Stage and commit.
5. Stage and commit.

This is certainly sorter than the above approach, but somehow I suspect, that I would continue to use the above approach anyway (but also start thinking harder about how navigating between "different representations of a change" could be made more convenient).

**Comment by npostavs** [comment]
> These are quite a few steps, but I doubt that the Ediff approach would take much less key strokes.

Ediff approach would be:

1. <kbd>e</kbd> on file which opens 3 buffers: working copy, index, and HEAD.
2. Make the change in the index buffer
3. <kbd>q</kbd> in the ediff control buffer to end ediff session, answer <kbd>y</kbd> to save index changes.
4. Commit (changes have been staged in step 3).

Is it that much less key strokes? Possibly not depending on how many key strokes making the change takes. But I still think it's significantly better than anything else.

**Comment by tarsius** [comment]
> Ediff approach would be:

That's actually quite nice, I think I'll go with that personally.

> Make the change in the index buffer

Which might begin with first staging the whole change (<kbd>c b</kbd>) and then *removing* the unwanted parts by editing the index (`B`) buffer.

@ Toxaris is that good enough for you? I might consider adding something like what I suggested above, but not before `v3.1`.

**Comment by Toxaris** [comment]
I’m aware that I could “edit the file, stage and commit, and then undo the changes”, but this is not an option for me because step 5 (“visit the file-visiting buffer”) and step 6 (“undo the changes”) rely on my memory. I want to focus on what I want to commit next, not on what I need to undo and in which buffers that is. Also philosophically, I feel that staging and commiting should not require me to change the working copy.

**Comment by tarsius** [comment]
@Toxaris what about the Ediff approach? I had some doubts about that myself, but after I tried it I actually like it.

> [don't want to] rely on my memory

With Ediff you have to do so even less. While you edit the index version you can still see the working tree version.

I recommend: `(setq ediff-window-setup-function 'ediff-setup-windows-plain)`.

**Comment by jsyjr** [comment]
I have just been experimenting this this ediff approach.  I very much like how it makes the index more tangible.

One wild idea: If I type `e` on a hunk (as opposed to a filename) could you bring up the ediff ui narrowed to the hunk in question?

**Comment by tarsius** [comment]
It probably should not narrow to the hunk, but it should jump to it just like `RET` does. I'll look into that once I get around to improve "jump to correct position" across all of magit.

**Comment by jsyjr** [comment]
I suggested the narrowing after confirming that I could indeed narrow the 3 ediff windows.  Not sure though how you might accomplish that as part of invoking ediff.

My motivation is that since magit already provides a "narrowed" view of the changes within a file I found it disconcerting to be dumped back into the context of the full file.  During my editing of the index I had to be careful with search and replace so that I modified only the relevant range.

Even if we quibble about whether to make it the default I very much would appreciate the option of being able to effect automatic narrowing.

PS: Thanks again for this most amazing tool.
